File: scripts/gen_dmg_bg.py
#!/usr/bin/env python3
"""Generate DMG installer background with ProgreSQL brand gradient.

Creates a 660x400 PNG with purple gradient (#6366f1 → #8b5cf6),
a curved arrow pointing from app icon to Applications folder,
and "ProgreSQL" text label.

Uses only stdlib modules (struct, zlib) — no Pillow required.
"""

import math
import struct
import zlib
import os

WIDTH = 660
HEIGHT = 400


def lerp_color(c1: tuple[int, int, int], c2: tuple[int, int, int], t: float) -> tuple[int, int, int]:
    return (
        int(c1[0] + (c2[0] - c1[0]) * t),
        int(c1[1] + (c2[1] - c1[1]) * t),
        int(c1[2] + (c2[2] - c1[2]) * t),
    )


def make_png(pixels: list[list[tuple[int, int, int, int]]], width: int, height: int) -> bytes:
    """Create a PNG file from RGBA pixel data."""

    def chunk(chunk_type: bytes, data: bytes) -> bytes:
        c = chunk_type + data
        return struct.pack(">I", len(data)) + c + struct.pack(">I", zlib.crc32(c) & 0xFFFFFFFF)

    header = b"\x89PNG\r\n\x1a\n"
    ihdr = chunk(b"IHDR", struct.pack(">IIBBBBB", width, height, 8, 6, 0, 0, 0))

    raw = b""
    for row in pixels:
        raw += b"\x00"  # filter byte
        for r, g, b, a in row:
            raw += struct.pack("BBBB", r, g, b, a)

    idat = chunk(b"IDAT", zlib.compress(raw, 9))
    iend = chunk(b"IEND", b"")
    return header + ihdr + idat + iend


def draw_gradient(pixels: list[list[tuple[int, int, int, int]]]) -> None:
    """Draw diagonal purple gradient."""
    c1 = (99, 102, 241)   # #6366f1
    c2 = (139, 92, 246)   # #8b5cf6
    c3 = (67, 56, 202)    # #4338ca — darker for depth

    for y in range(HEIGHT):
        for x in range(WIDTH):
            # Diagonal gradient: top-left → bottom-right
            t = (x / WIDTH * 0.6 + y / HEIGHT * 0.4)
            # Mix c3→c1→c2 for richer gradient
            if t < 0.5:
                color = lerp_color(c3, c1, t * 2)
            else:
                color = lerp_color(c1, c2, (t - 0.5) * 2)
            pixels[y][x] = (color[0], color[1], color[2], 255)


def draw_circle_aa(
    pixels: list[list[tuple[int, int, int, int]]],
    cx: float, cy: float, r: float,
    color: tuple[int, int, int], alpha: int = 255
) -> None:
    """Draw a filled circle with basic anti-aliasing."""
    for y in range(max(0, int(cy - r - 2)), min(HEIGHT, int(cy + r + 2))):
        for x in range(max(0, int(cx - r - 2)), min(WIDTH, int(cx + r + 2))):
            dist = math.sqrt((x - cx) ** 2 + (y - cy) ** 2)
            if dist <= r - 0.5:
                a = alpha
            elif dist <= r + 0.5:
                a = int(alpha * (r + 0.5 - dist))
            else:
                continue
            bg = pixels[y][x]
            t = a / 255.0
            pixels[y][x] = (
                int(bg[0] * (1 - t) + color[0] * t),
                int(bg[1] * (1 - t) + color[1] * t),
                int(bg[2] * (1 - t) + color[2] * t),
                255,
            )


def draw_thick_line(
    pixels: list[list[tuple[int, int, int, int]]],
    x1: float, y1: float, x2: float, y2: float,
    thickness: float,
    color: tuple[int, int, int], alpha: int = 255
) -> None:
    """Draw an anti-aliased thick line."""
    dx = x2 - x1
    dy = y2 - y1
    length = math.sqrt(dx * dx + dy * dy)
    if length == 0:
        return

    # Normal vector
    nx = -dy / length
    ny = dx / length

    min_x = max(0, int(min(x1, x2) - thickness - 1))
    max_x = min(WIDTH, int(max(x1, x2) + thickness + 2))
    min_y = max(0, int(min(y1, y2) - thickness - 1))
    max_y = min(HEIGHT, int(max(y1, y2) + thickness + 2))

    half = thickness / 2.0

    for y in range(min_y, max_y):
        for x in range(min_x, max_x):
            # Project point onto line
            px, py = x - x1, y - y1
            along = (px * dx + py * dy) / length
            across = abs(px * nx + py * ny)

            if along < -half or along > length + half:
                continue

            # Distance to line segment
            if along < 0:
                dist = math.sqrt(px * px + py * py)
            elif along > length:
                qx, qy = x - x2, y - y2
                dist = math.sqrt(qx * qx + qy * qy)
            else:
                dist = across

            if dist <= half - 0.5:
                a = alpha
            elif dist <= half + 0.5:
                a = int(alpha * (half + 0.5 - dist))
            else:
                continue

            bg = pixels[y][x]
            t = a / 255.0
            pixels[y][x] = (
                int(bg[0] * (1 - t) + color[0] * t),
                int(bg[1] * (1 - t) + color[1] * t),
                int(bg[2] * (1 - t) + color[2] * t),
                255,
            )


def draw_curved_arrow(pixels: list[list[tuple[int, int, int, int]]]) -> None:
    """Draw a curved arrow from app icon area to Applications folder area."""
    color = (220, 220, 255)  # Light lavender
    thickness = 3.5

    # Arc from (200, 230) to (460, 230) with peak at y=140
    cx_arc = 330
    cy_arc = 330
    r_arc = 160
    start_angle = math.radians(200)
    end_angle = math.radians(340)

    steps = 80
    prev_x, prev_y = None, None
    end_x, end_y = 0.0, 0.0

    for i in range(steps + 1):
        t = i / steps
        angle = start_angle + (end_angle - start_angle) * t
        ax = cx_arc + r_arc * math.cos(angle)
        ay = cy_arc + r_arc * math.sin(angle)
        if prev_x is not None:
            draw_thick_line(pixels, prev_x, prev_y, ax, ay, thickness, color, 180)
        prev_x, prev_y = ax, ay
        end_x, end_y = ax, ay

    # Arrowhead at the end
    arrow_angle = end_angle + math.radians(10)
    arrow_len = 18
    a1_angle = arrow_angle + math.radians(150)
    a2_angle = arrow_angle - math.radians(150)
    a1x = end_x + arrow_len * math.cos(a1_angle)
    a1y = end_y + arrow_len * math.sin(a1_angle)
    a2x = end_x + arrow_len * math.cos(a2_angle)
    a2y = end_y + arrow_len * math.sin(a2_angle)

    draw_thick_line(pixels, end_x, end_y, a1x, a1y, thickness, color, 180)
    draw_thick_line(pixels, end_x, end_y, a2x, a2y, thickness, color, 180)


# Simple bitmap font for "ProgreSQL" text
FONT_GLYPHS: dict[str, list[str]] = {
    "P": [
        "####.",
        "#...#",
        "####.",
        "#....",
        "#....",
    ],
    "r": [
        ".....",
        "#.##.",
        "##...",
        "#....",
        "#....",
    ],
    "o": [
        ".....",
        ".##..",
        "#..#.",
        "#..#.",
        ".##..",
    ],
    "g": [
        ".....",
        ".###.",
        "#..#.",
        ".###.",
        "...#.",
    ],
    "e": [
        ".....",
        ".##..",
        "####.",
        "#....",
        ".##..",
    ],
    "S": [
        ".###.",
        "#....",
        ".##..",
        "...#.",
        "###..",
    ],
    "Q": [
        ".##..",
        "#..#.",
        "#..#.",
        "#.#..",
        ".##.",
    ],
    "L": [
        "#....",
        "#....",
        "#....",
        "#....",
        "####.",
    ],
    "s": [
        ".....",
        ".##..",
        "#....",
        ".##..",
        "##...",
    ],
    "q": [
        ".....",
        ".###.",
        "#..#.",
        ".###.",
        "...#.",
    ],
}


def draw_text(
    pixels: list[list[tuple[int, int, int, int]]],
    text: str, start_x: int, start_y: int,
    scale: int, color: tuple[int, int, int], alpha: int = 200
) -> None:
    """Render text using simple bitmap font."""
    cursor_x = start_x
    for ch in text:
        glyph = FONT_GLYPHS.get(ch)
        if glyph is None:
            cursor_x += 3 * scale  # space
            continue
        for gy, row in enumerate(glyph):
            for gx, pixel in enumerate(row):
                if pixel == "#":
                    for dy in range(scale):
                        for dx in range(scale):
                            px = cursor_x + gx * scale + dx
                            py = start_y + gy * scale + dy
                            if 0 <= px < WIDTH and 0 <= py < HEIGHT:
                                bg = pixels[py][px]
                                t = alpha / 255.0
                                pixels[py][px] = (
                                    int(bg[0] * (1 - t) + color[0] * t),
                                    int(bg[1] * (1 - t) + color[1] * t),
                                    int(bg[2] * (1 - t) + color[2] * t),
                                    255,
                                )
        cursor_x += (len(glyph[0]) + 1) * scale


def draw_subtle_dots(pixels: list[list[tuple[int, int, int, int]]]) -> None:
    """Draw subtle decorative dots/stars pattern."""
    import random
    random.seed(42)  # deterministic
    for _ in range(60):
        x = random.randint(0, WIDTH - 1)
        y = random.randint(0, HEIGHT - 1)
        size = random.uniform(0.5, 1.5)
        alpha = random.randint(30, 80)
        draw_circle_aa(pixels, x, y, size, (255, 255, 255), alpha)


def main() -> None:
    # Initialize pixel buffer
    pixels: list[list[tuple[int, int, int, int]]] = [
        [(0, 0, 0, 255)] * WIDTH for _ in range(HEIGHT)
    ]

    # 1. Purple gradient background
    draw_gradient(pixels)

    # 2. Subtle decorative dots
    draw_subtle_dots(pixels)

    # 3. Curved arrow
    draw_curved_arrow(pixels)

    # 4. "ProgreSQL" text at bottom
    text_color = (255, 255, 255)
    text = "ProgreSQL"
    scale = 3
    char_width = 6 * scale  # ~18px per char
    text_width = len(text) * char_width
    text_x = (WIDTH - text_width) // 2
    text_y = HEIGHT - 60
    draw_text(pixels, text, text_x, text_y, scale, text_color, 200)

    # Generate PNG
    png_data = make_png(pixels, WIDTH, HEIGHT)

    # Write files
    out_dir = os.path.join(os.path.dirname(__file__), "..", "frontend", "public", "assets", "dmg")
    os.makedirs(out_dir, exist_ok=True)

    out_path = os.path.join(out_dir, "background.png")
    with open(out_path, "wb") as f:
        f.write(png_data)
    print(f"Written: {out_path} ({len(png_data)} bytes)")

    # Also generate @2x version (1320x800)
    # For now, just copy the same since create-dmg handles scaling
    out_path_2x = os.path.join(out_dir, "background@2x.png")
    with open(out_path_2x, "wb") as f:
        f.write(png_data)
    print(f"Written: {out_path_2x} ({len(png_data)} bytes)")


if __name__ == "__main__":
    main()

File: scripts/gen_dmg_bg.js
#!/usr/bin/env node
/**
 * Generate DMG installer background with ProgreSQL brand gradient.
 * Creates a 660x400 PNG with purple gradient (#6366f1 -> #8b5cf6),
 * a curved arrow, and "ProgreSQL" text.
 * Pure Node.js — no external dependencies.
 */

const fs = require("fs");
const zlib = require("zlib");

const WIDTH = 660,
  HEIGHT = 400;

function crc32(buf) {
  let crc = 0xffffffff;
  for (let i = 0; i < buf.length; i++) {
    crc ^= buf[i];
    for (let j = 0; j < 8; j++) {
      crc = (crc >>> 1) ^ (crc & 1 ? 0xedb88320 : 0);
    }
  }
  return (crc ^ 0xffffffff) >>> 0;
}

function chunk(type, data) {
  const len = Buffer.alloc(4);
  len.writeUInt32BE(data.length);
  const td = Buffer.concat([Buffer.from(type), data]);
  const crc = Buffer.alloc(4);
  crc.writeUInt32BE(crc32(td));
  return Buffer.concat([len, td, crc]);
}

function lerp(a, b, t) {
  return Math.round(a + (b - a) * t);
}

// Build pixel data with gradient
const raw = Buffer.alloc(HEIGHT * (1 + WIDTH * 4));
const c1 = [67, 56, 202]; // #4338ca
const c2 = [99, 102, 241]; // #6366f1
const c3 = [139, 92, 246]; // #8b5cf6

// Pseudo-random dots
let rng = 42;
function nextRng() {
  rng = (rng * 1103515245 + 12345) & 0x7fffffff;
  return rng / 0x7fffffff;
}

const dots = [];
for (let i = 0; i < 60; i++) {
  dots.push({
    x: Math.floor(nextRng() * WIDTH),
    y: Math.floor(nextRng() * HEIGHT),
    a: 30 + Math.floor(nextRng() * 50),
  });
}

for (let y = 0; y < HEIGHT; y++) {
  const rowOff = y * (1 + WIDTH * 4);
  raw[rowOff] = 0;
  for (let x = 0; x < WIDTH; x++) {
    const t = x / WIDTH * 0.6 + y / HEIGHT * 0.4;
    let r, g, b;
    if (t < 0.5) {
      r = lerp(c1[0], c2[0], t * 2);
      g = lerp(c1[1], c2[1], t * 2);
      b = lerp(c1[2], c2[2], t * 2);
    } else {
      r = lerp(c2[0], c3[0], (t - 0.5) * 2);
      g = lerp(c2[1], c3[1], (t - 0.5) * 2);
      b = lerp(c2[2], c3[2], (t - 0.5) * 2);
    }

    for (const d of dots) {
      const dist = Math.sqrt((x - d.x) ** 2 + (y - d.y) ** 2);
      if (dist < 1.5) {
        const da = (d.a * Math.max(0, 1 - dist / 1.5)) / 255;
        r = Math.round(r * (1 - da) + 255 * da);
        g = Math.round(g * (1 - da) + 255 * da);
        b = Math.round(b * (1 - da) + 255 * da);
      }
    }

    const off = rowOff + 1 + x * 4;
    raw[off] = r;
    raw[off + 1] = g;
    raw[off + 2] = b;
    raw[off + 3] = 255;
  }
}

function blendPixel(x, y, r, g, b, a) {
  x = Math.round(x);
  y = Math.round(y);
  if (x < 0 || x >= WIDTH || y < 0 || y >= HEIGHT) return;
  const off = y * (1 + WIDTH * 4) + 1 + x * 4;
  const t = a / 255;
  raw[off] = Math.round(raw[off] * (1 - t) + r * t);
  raw[off + 1] = Math.round(raw[off + 1] * (1 - t) + g * t);
  raw[off + 2] = Math.round(raw[off + 2] * (1 - t) + b * t);
}

// Draw curved arrow
const arrowColor = [220, 220, 255];
const cx = 330, cy = 330, rad = 160;
const startA = (200 * Math.PI) / 180;
const endA = (340 * Math.PI) / 180;
const thickness = 3.5;

let lastX, lastY;
for (let i = 0; i <= 200; i++) {
  const angle = startA + (endA - startA) * (i / 200);
  const ax = cx + rad * Math.cos(angle);
  const ay = cy + rad * Math.sin(angle);
  if (i > 0) {
    const dx = ax - lastX,
      dy = ay - lastY;
    const len = Math.sqrt(dx * dx + dy * dy);
    if (len > 0) {
      const nx = -dy / len,
        ny = dx / len;
      const minX = Math.max(0, Math.floor(Math.min(lastX, ax) - thickness - 1));
      const maxX = Math.min(WIDTH - 1, Math.ceil(Math.max(lastX, ax) + thickness + 1));
      const minY = Math.max(0, Math.floor(Math.min(lastY, ay) - thickness - 1));
      const maxY = Math.min(HEIGHT - 1, Math.ceil(Math.max(lastY, ay) + thickness + 1));
      const half = thickness / 2;
      for (let py = minY; py <= maxY; py++) {
        for (let px = minX; px <= maxX; px++) {
          const ppx = px - lastX,
            ppy = py - lastY;
          const along = (ppx * dx + ppy * dy) / len;
          if (along < -half || along > len + half) continue;
          let dist;
          if (along < 0) dist = Math.sqrt(ppx * ppx + ppy * ppy);
          else if (along > len) {
            const qx = px - ax,
              qy = py - ay;
            dist = Math.sqrt(qx * qx + qy * qy);
          } else dist = Math.abs(ppx * nx + ppy * ny);
          if (dist <= half + 0.5) {
            const a = dist <= half - 0.5 ? 180 : Math.round(180 * (half + 0.5 - dist));
            blendPixel(px, py, arrowColor[0], arrowColor[1], arrowColor[2], a);
          }
        }
      }
    }
  }
  lastX = ax;
  lastY = ay;
}

// Arrowhead
const endX = cx + rad * Math.cos(endA);
const endY = cy + rad * Math.sin(endA);
const arrowAngle = endA + (10 * Math.PI) / 180;
const arrowLen = 18;
for (const offset of [150, -150]) {
  const a = arrowAngle + (offset * Math.PI) / 180;
  const tx = endX + arrowLen * Math.cos(a);
  const ty = endY + arrowLen * Math.sin(a);
  for (let i = 0; i < 30; i++) {
    const t = i / 30;
    const px = endX + (tx - endX) * t;
    const py = endY + (ty - endY) * t;
    for (let dy = -2; dy <= 2; dy++)
      for (let dx = -2; dx <= 2; dx++) {
        const d = Math.sqrt(dx * dx + dy * dy);
        if (d <= 2)
          blendPixel(
            px + dx,
            py + dy,
            arrowColor[0],
            arrowColor[1],
            arrowColor[2],
            Math.round(180 * Math.max(0, 1 - d / 2)),
          );
      }
  }
}

// Draw "ProgreSQL" text
const glyphs = {
  P: ["####.", "#...#", "####.", "#....", "#...."],
  r: [".....", "#.##.", "##...", "#....", "#...."],
  o: [".....", ".##..", "#..#.", "#..#.", ".##.."],
  g: [".....", ".###.", "#..#.", ".###.", "...#."],
  e: [".....", ".##..", "####.", "#....", ".##.."],
  S: [".###.", "#....", ".##..", "...#.", "###.."],
  Q: [".##..", "#..#.", "#..#.", "#.#..", ".##.#"],
  L: ["#....", "#....", "#....", "#....", "####."],
};
const text = "ProgreSQL";
const scale = 3;
const charW = 6 * scale;
const textW = text.length * charW;
const textX = Math.floor((WIDTH - textW) / 2);
const textY = HEIGHT - 60;

for (let ci = 0; ci < text.length; ci++) {
  const gl = glyphs[text[ci]];
  if (!gl) continue;
  for (let gy = 0; gy < gl.length; gy++) {
    for (let gx = 0; gx < gl[gy].length; gx++) {
      if (gl[gy][gx] === "#") {
        for (let dy = 0; dy < scale; dy++)
          for (let dx = 0; dx < scale; dx++) {
            blendPixel(textX + ci * charW + gx * scale + dx, textY + gy * scale + dy, 255, 255, 255, 200);
          }
      }
    }
  }
}

// Create PNG
const sig = Buffer.from([137, 80, 78, 71, 13, 10, 26, 10]);
const ihdrData = Buffer.alloc(13);
ihdrData.writeUInt32BE(WIDTH, 0);
ihdrData.writeUInt32BE(HEIGHT, 4);
ihdrData[8] = 8;
ihdrData[9] = 6; // 8-bit RGBA
const ihdr = chunk("IHDR", ihdrData);
const compressed = zlib.deflateSync(raw, { level: 9 });
const idat = chunk("IDAT", compressed);
const iend = chunk("IEND", Buffer.alloc(0));
const png = Buffer.concat([sig, ihdr, idat, iend]);

const outDir = "frontend/public/assets/dmg";
fs.writeFileSync(outDir + "/background.png", png);
fs.writeFileSync(outDir + "/background@2x.png", png);
console.log("Written: " + outDir + "/background.png (" + png.length + " bytes)");
console.log("Written: " + outDir + "/background@2x.png (" + png.length + " bytes)");
